File: articles/calculus-for-systems-modeling/articles/inverse-functions-and-system-interpretation/r/inverse_interpretation_audit.R
forward_model <- function(x) {
  log1p(x)
}

forward_derivative <- function(x) {
  1 / (1 + x)
}

inverse_model <- function(y) {
  exp(y) - 1
}

inverse_audit <- function(target_output) {
  x <- inverse_model(target_output)
  y_check <- forward_model(x)
  residual <- y_check - target_output
  derivative <- forward_derivative(x)
  inverse_sensitivity <- 1 / derivative
  domain_valid <- x > -1

  warning <- ""
  if (!domain_valid) {
    warning <- "recovered input outside domain"
  } else if (abs(derivative) < 1e-6) {
    warning <- "small forward derivative; inverse may be unstable"
  } else if (abs(residual) > 1e-8) {
    warning <- "forward check does not reproduce target output"
  }

  data.frame(
    target_output = target_output,
    recovered_input = x,
    forward_check = y_check,
    residual = residual,
    forward_derivative = derivative,
    inverse_sensitivity = inverse_sensitivity,
    domain_valid = domain_valid,
    warning = warning
  )
}

results <- do.call(rbind, lapply(c(0, 0.5, 1, 1.5, 2), inverse_audit))
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_inverse_interpretation_audit.csv", row.names = FALSE)
print(results)

File: articles/calculus-for-systems-modeling/articles/inverse-functions-and-system-interpretation/fortran/inverse_interpretation_audit.f90
program inverse_interpretation_audit
  implicit none
  real(8), dimension(5) :: ys
  real(8) :: y,x,ycheck,residual,derivative,invsens
  integer :: i
  ys = (/0.0d0,0.5d0,1.0d0,1.5d0,2.0d0/)
  print '(A)', 'target_output recovered_input forward_check residual forward_derivative inverse_sensitivity'
  do i=1,size(ys)
    y=ys(i); x=inverse_model(y); ycheck=forward_model(x); residual=ycheck-y
    derivative=forward_derivative(x); invsens=1.0d0/derivative
    print '(F8.3,1X,F14.6,1X,F14.6,1X,F14.6,1X,F14.6,1X,F14.6)', y,x,ycheck,residual,derivative,invsens
  end do
contains
  real(8) function forward_model(x)
    real(8), intent(in) :: x
    forward_model=log(1.0d0+x)
  end function
  real(8) function forward_derivative(x)
    real(8), intent(in) :: x
    forward_derivative=1.0d0/(1.0d0+x)
  end function
  real(8) function inverse_model(y)
    real(8), intent(in) :: y
    inverse_model=exp(y)-1.0d0
  end function
end program
